### PR TITLE
Use styled-components

### PR DESCRIPTION
**Pull request by samit4me**
### Styled Components #1100

Modified containers and components to make use of the awesome new [styled-compoents](https://github.com/styled-components/styled-components). Essentially this means swapping `styles.css` files for Styled Components.

I have done the bare minimum to introduce `styled-components` and super keen to hear any feedback (don't hold back). During this change, I also noticed a few things and would like to hear any thoughts on the following:
- Use stateless functions where possible _(some components have ESLint comments disabling the warnings)_.
- Use arrow functions instead of regular functions for stateless components _(debatably cleaner syntax)_.
- Module import ordering, it is currently a mix match, but the theme I see mostly is, 3rd party first, followed by a new line then project modules, all in alphanumeric order.

**Note:** In many instances I needed to switch from Enzyme's `shallow` to either `mount` or `render`. This is because there is no way _(that I know of)_ to find a specific StyledComponent with shallow rendering as they are not a React constructor and they all have the same dislayName `StyledComponent`.

**Important:** This may be a breaking change as the `<A>`, `<Button>`, `<ListItem>` components will no longer allow the `className` prop for overriding the styling. If you are using this feature I would recommend looking at how to override styles in `styled-components`, see [Overriding component styles](https://github.com/styled-components/styled-components#overriding-component-styles).

**Optional:** Due to the nature of `styled-components`, generating a `styles.css` file inside of each component is mostly not required. Should we update the component generator code so it no longer prompts for style files? Or leave this as is for now?

### CODE REVIEW

**Comment by mxstbr** [comment]
> Due to the nature of styled-components, generating a styles.css file inside of each component is mostly not required.

We should not have any CSS files anymore, remove them from the scaffolding and remove the support for CSS modules from the webpack configuration!

**Comment by sedubois** [comment]
> Use stateless functions where possible

I would also prefer stateless functions for their conciseness and to ensure all state is handled within Redux, but really recommend against doing so as currently prevents HMR. I would actually recommend either disabling the react/prefer-stateless-function altogether or just continue to add it on every class component, until HMR is supported in functional components.

**Comment by sedubois** [comment]
> there is no way (that I know of) to find a specific StyledComponent with shallow rendering as they are not a React constructor and they all have the same dislayName StyledComponent

That sounds like a problem that should be fixed by styled-components instead of being workarounded by getting rid of `shallow`?

File: app/components/A/tests/index.test.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ describe('<A />', () => {
     expect(renderedComponent.type()).toEqual('a');
   });
 
-  it('should have an href attribute', () => {
+  it('should have a href attribute', () => {
```
Comment (sedubois) [inline app/components/A/tests/index.test.js]: can revert this change 😉

File: app/components/Button/index.js [hunk 2]
```diff
@@ -7,37 +7,64 @@
...
+  color: #41ADDD;
 
-function Button(props) {
-  const className = props.className ? props.className : styles.button;
+  &:active {
+    background: #41ADDD;
+    color: #FFF;
+  }
+`;
+export const AnchorButton = styled.a`${buttonStyling}`;
```
Comment (sedubois) [inline app/components/Button/index.js]: I don't see why this is called a button as it's an anchor. Maybe name `A` or `Link`?

**Comment by samit4me** [comment]
> We should not have any CSS files anymore, remove them from the scaffolding and remove the support for CSS modules from the webpack configuration!

If we remove all CSS files how will we apply Global styles? In particular, styles for elements outside of the React container (e.g. `<div id="app"></div>`)? Did you want to inline those styles into `index.html`?

File: app/components/A/tests/index.test.js (hunk 1 above)
Comment (samit4me) [inline app/components/A/tests/index.test.js]: Pretty much all other tests regarding attributes in this codebase are worded in this way *(e.g. 'should have a ? attribute')* and this is why I changed this one to match. IMO **a** also reads better than **an** in this instance. I can change it back if you think it should be **an**, but does that mean all the others should also be updated?

**Comment by GGAlanSmithee** [comment]
@sedubois

> to ensure all state is handled within Redux

I think a more proper way to word it would be for all *global* state to be handled by redux. If a component is (locally) stateful it should have a local state, which should not be handled by Redux (IMO).

> recommend against doing so as currently prevents HMR

I (sadly) agree with this statement. Not being able to hot reload stateless functions is super disruptive to an otherwise awesome developer experience. It actually looks like react hot loaded 3 might be released in not to distant future, which should remove this constraint if I understand it correctly.

File: app/components/Button/index.js (hunk 2 above)
Comment (samit4me) [inline app/components/Button/index.js]: Your totally right, it could be called `A` or `Link` and that makes sense. I simply choose `AnchorButton` because I thought it was more descriptive.

**Comment by mxstbr** [comment]
> If we remove all CSS files how will we apply Global styles?

styled-components has an `injectGlobal` helper that allows you to do that! Add a `global-styles.js` file in the `app/` folder and put the global styles there!

**Comment by samit4me** [comment]
> styled-components has an injectGlobal helper that allows you to do that!

That is fantastic thank you, not sure how I missed that one. Is there a way to lint or minimize the css in `styled-components` Or would that be a feature request? I'm guessing something like that would have to be done using a babel plugin.

**Comment by mxstbr** [comment]
> Is there a way to lint or minimize the css in styled-components

Not yet, but that's definitely one of the next steps!

**Comment by diegohaz** [comment]
I think this PR should be used as a migration reference when it's done. Helped me a lot.

**Comment by mxstbr** [comment]
Sorry, I merged a PR by @sedubois because it looked so great without realising the implications for this ongoing work – would you mind rebasing on top of the recent changes? Sorry for that, my bad!

**Comment by samit4me** [comment]
> Sorry, I merged a PR by @sedubois because it looked so great without realising the implications for this ongoing work – would you mind rebasing on top of the recent changes? Sorry for that, my bad!

No worries mate, I'll have a quick look now, but leaving for work shortly so it might have to wait until tonight if there are too many conflicts.

**Comment by samit4me** [comment]
@mxstbr I am new to **rebase** but I gave it a try, which seems to work fine, but I had to force push to my branch. Is that okay, will this cause any issues?

**Comment by sedubois** [comment]
@samit4me that's perfectly fine, I use rebase + push --force quite regularly. Push --force to a dev branch that you own yourself is fine, especially if you know nobody has pulled it to add changes on top. I like rebase as it keeps things quite clean. But if it scares you, don't hesitate another time to just merge and when merging to `dev`, we can just squash.

**Comment by mxstbr** [comment]
@samit4me that sounds perfectly right!

**Comment by samit4me** [comment]
Great thanks @sedubois @mxstbr . I too prefer to rebase (now), it's much cleaner. I told a colleague about it and he said I was doing it wrong (called me evil), so I got paranoid. It does make sense to push -f onto your own branch that nobody else has pulled. Cheers guys :)

**Comment by GGAlanSmithee** [comment]
@samit4me he probably had some bad experience with someone rebasing and forcepushing a branch that he was also working on, but what you did is the correct way to do it 👍

**Comment by mxstbr** [comment]
Before I dive into the code, is this ready to be reviewed?

**Comment by samit4me** [comment]
> Before I dive into the code, is this ready to be reviewed?

Yep go for it.

**Comment by mxstbr** [comment]
Reviewing this now. Looking at the styling everything looks the exact same – we're off to a good start 😎

File: README.md [hunk 3]
```diff
@@ -93,7 +93,7 @@ But wait... there's more!
 
 There’s also a <a href="https://vimeo.com/168648012">fantastic video</a> on how to structure your React.js apps with scalability in mind. It provides rationale for the majority of boilerplate's design decisions.
 
-<sub><i>Keywords: React.js, Redux, Hot Reloading, ESNext, Babel, PostCSS, Autoprefixer, react-router, Offline First, ServiceWorker, CSS Modules, redux-saga, FontFaceObserver, PageSpeed Insights</i></sub>
+<sub><i>Keywords: React.js, Redux, Hot Reloading, ESNext, Babel, react-router, Offline First, ServiceWorker, redux-saga, FontFaceObserver, PageSpeed Insights</i></sub>
```
Comment (mxstbr) [inline README.md]: Let's mention `styled-components` here!

File: app/components/A/tests/index.test.js (hunk 1 above)
Comment (mxstbr) [inline app/components/A/tests/index.test.js]: `h` in english is treated like a vocal at the beginning of words. You say "AN apple" but you say "A banana" because apple starts with a vocal – and `h` is treated like a vocal at the beginning of words, so this should be "AN href"

File: app/components/ListItem/index.js [hunk 4]
```diff
@@ -1,19 +1,39 @@
...
+
+const Item = styled.li`
+  width: 100%;
+  height: 3em;
+  display: flex;
+  align-items: center;
+  position: relative;
+  border-top: 1px solid #EEE;
+
+  &&:first-child {
```
Comment (mxstbr) [inline app/components/ListItem/index.js]: Unnecessary double ampersand? I think `&:first-child` should work just fine?

File: docs/README.md [hunk 5]
```diff
@@ -15,8 +15,7 @@
   - [Component Testing](testing/component-testing.md)
   - [Remote Testing](testing/remote-testing.md)
 - [CSS](css)
-  - [PostCSS](css/postcss.md)
-  - [CSS Modules](css/css-modules.md)
+  - [Styled Components](css/styled-componets.md)
```
Comment (mxstbr) [inline docs/README.md]: nit: write `styled-components` (including the code block!)

File: docs/css/README.md [hunk 6]
```diff
@@ -1,15 +1,13 @@
 # CSS
 
-This boilerplate uses PostCSS as a CSS preprocessor with a few utility plugins
-to make it "batteries included".
+This boilerplate uses Styled Components allowing you to write your CSS in your JavaScript,
```
Comment (mxstbr) [inline docs/css/README.md]: nit: write `styled-components` (including the code block!) and link to the s-c repo

File: docs/css/README.md [hunk 7]
```diff
@@ -1,15 +1,13 @@
...
+removing the mapping between styles and components.
 
-CSS Modules lets us embrace component encapsulation while sanitize.css gives us
+Styled Components lets us embrace component encapsulation while sanitize.css gives us
```
Comment (mxstbr) [inline docs/css/README.md]: nit: write `styled-components` (including the code block!), let's (missing ')

File: docs/css/README.md [hunk 8]
```diff
@@ -1,15 +1,13 @@
...
 data-driven cross-browser normalisation.
 
 Learn more:
 
-- [PostCSS](postcss.md)
-- [CSS Modules](css-modules.md)
+- [Styled Components](styled-componets.md)
```
Comment (mxstbr) [inline docs/css/README.md]: nit: write `styled-components` (including the code block!)

File: docs/css/sass.md [hunk 9]
```diff
@@ -1,42 +1,21 @@
 # Can I use Sass with this boilerplate?
 
 Yes, although we advise against it and **do not support this**. We selected
-PostCSS over Sass because its approach is more powerful: instead of trying to
+[Styled Components](https://github.com/styled-components/styled-components)
```
Comment (mxstbr) [inline docs/css/sass.md]: nit: write `styled-components` (including the code block!)

File: docs/css/styled-componets.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
+# Styled Components
```
Comment (mxstbr) [inline docs/css/styled-componets.md]: nit: write `styled-components` (including the code block!)

File: docs/css/styled-componets.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+Styled Components allow you to write actual CSS code in your JavaScript to style your components,
```
Comment (mxstbr) [inline docs/css/styled-componets.md]: nit: write `styled-components` (including the code block!)

File: docs/general/README.md [hunk 12]
```diff
@@ -54,12 +54,12 @@ and `react` presets!
 
 ## Next generation CSS
 
-Write composable CSS that's co-located with your components using [CSS modules]
+Write composable CSS that's co-located with your components using [Styled Components]
```
Comment (mxstbr) [inline docs/general/README.md]: nit: write `styled-components` (including the code block!)

File: docs/general/README.md [hunk 13]
```diff
@@ -54,12 +54,12 @@ and `react` presets!
...
 for complete modularity. Unique generated class names keep the specificity low
 while eliminating style clashes. Ship only the styles that are used on the
 visible page for the best performance.
 
-[CSS modules]: ../css/css-modules.md
+[Styled Components]: ../css/styled-components.md
```
Comment (mxstbr) [inline docs/general/README.md]: nit: write `styled-components` (including the code block!)

File: docs/general/faq.md [hunk 14]
```diff
@@ -1,6 +1,6 @@
 # Frequently Asked Questions
 
-## Where are Babel, ESLint and stylelint configured?
+## Where are Babel are ESLint configured?
```
Comment (mxstbr) [inline docs/general/faq.md]: nit: typo, "are" should be "and"

File: app/components/Button/index.js [hunk 15]
```diff
@@ -7,37 +7,64 @@
  */
 
 import React, { PropTypes, Children } from 'react';
+import styled from 'styled-components';
 
-import styles from './styles.css';
+const buttonStyling = `
```
Comment (mxstbr) [inline app/components/Button/index.js]: Use `css` for the `buttonStyles` like so:

  ```JS
  import styled, { css } from 'styled-components';

  const buttonStyles = css`
    styles: here;
  `;
  ```

Otherwise if people add interpolated functions here they won't work!

**Comment by mxstbr** [comment]
> The only real issues I had was testing

That's great feedback! We have an issue open about testing, [feel free to chime in](https://github.com/styled-components/styled-components/issues/118) with what you think it should look like, trying to figure that out!

Minification is a good point and something I hadn't thought of. I'm working on linting as we speak (and am quite close), but minification is definitely a thing I hadn't considered. Might need to add `styled-components` support to Uglify/Babili for that to work…

File: app/components/Button/buttonStyle.js [hunk 16]
```diff
@@ -1,10 +1,4 @@
-.buttonWrapper {
-  width: 100%;
-  text-align: center;
-  margin: 4em 0;
-}
-
-.button {
+const buttonStyles = `
```
Comment (mxstbr) [inline app/components/Button/buttonStyle.js]: Please add `css` here:

```JS
import { css } from 'styled-components';

const buttonStyles = css`
  styles: here;
`;
```

**Comment by mxstbr** [comment]
Wow this looks great, amazing job! One tiny nitpick and then we can :ship: this!!

**Comment by samit4me** [comment]
@mxstbr Thank you, good find, I totally missed the `css` helper function 😳

**Comment by samit4me** [comment]
@mxstbr Should we add your new linter in with this PR? It would make sense as we removed the previous linter and this way nothing is lost!

**Comment by mxstbr** [comment]
Please try it – I'm unsure if it works yet, but this'll be a great opportunity to squash last bugs!

**Comment by samit4me** [comment]
@mxstbr I tried to include the linter but ran into issues around indentation and interpolation. I checked out the linter source code and can see the passing tests, so not sure what this issue is, but I am assuming it is me. I am not sure if you want to wait until these issues are sorted or merge as is without the linter???

The linter errors I am getting are listed below.
```
app/components/Button/A.js
 5:11  ×  Unexpected empty block   block-no-empty

app/components/Button/StyledButton.js
 5:11  ×  Unexpected empty block   block-no-empty

app/components/LoadingIndicator/Circle.js
 17:1   ×  Expected empty line before non-nested rule                                                 rule-non-nested-empty-line-before
 18:5   ×  Expected indentation of 2 spaces                                                           indentation
 19:5   ×  Expected indentation of 2 spaces                                                           indentation
 20:5   ×  Expected indentation of 2 spaces                                                           indentation
 21:5   ×  Expected indentation of 2 spaces                                                           indentation
 22:5   ×  Expected indentation of 2 spaces                                                           indentation
 25:1   ×  Expected no more than 1 empty line(s)                                                      max-empty-lines
 25:5   ×  Expected indentation of 2 spaces                                                           indentation
 25:6   ×  Expected double colon pseudo-element notation                                              selector-pseudo-element-colon-notation
 26:7   ×  Expected indentation of 4 spaces                                                           indentation
 27:7   ×  Expected indentation of 4 spaces                                                           indentation
 28:7   ×  Expected indentation of 4 spaces                                                           indentation
 29:7   ×  Expected indentation of 4 spaces                                                           indentation
 30:7   ×  Expected indentation of 4 spaces                                                           indentation
 31:7   ×  Expected indentation of 4 spaces                                                           indentation
 32:7   ×  Expected indentation of 4 spaces                                                           indentation
 32:32  ×  Unexpected whitespace at end of line                                                       no-eol-whitespace
 33:4   ×  Expected newline before "}" of a multi-line block (block-closing-backtick-newline-before)  block-closing-brace-newline-before
 33:5   ×  Expected newline before "}" of a multi-line block (block-closing-backtick-newline-before)  block-closing-brace-newline-before
 33:5   ×  Expected indentation of 2 spaces                                                           indentation
 33:6   ×  Unexpected empty line before closing backtick                                              block-closing-brace-empty-line-before
``

**Comment by mxstbr** [comment]
> but I am assuming it is me

Probably isn't if the source code looks correct 😉 That's why I wanted you to try! Can you try adding some failing tests with copy-and-paste of broken things to the stylelint processor?

**Comment by mxstbr** [comment]
I'm keen to ship this, even without the linter. I'll rebase this and merge it into the new version!

**Comment by mxstbr** [comment]
Did you remove the linter in that last commit?

**Comment by samit4me** [comment]
Oh boy, literally, the moment I hit enter on the `git push` I heard a loud bang outside and dogs barking, I race out and it looks as tho a possum has fallen out of the tree and our two dogs were chasing it around like cat and mouse, knocking over furniture and what not. So I think it is safe to say this PR is going out with a BANG! (terrible I know). In case you were wondering, I manage to save the possum from the dogs and relocated it to another tree out the front 👍 🇦🇺

**Comment by samit4me** [comment]
@mxstbr The linter was never committed, was only experimenting with it locally. The previous css linter has also been removed.

**Comment by mxstbr** [comment]
> Oh boy, literally, the moment I hit enter on the git push I heard a loud bang outside and dogs barking, I race out and it looks as tho a possum has fallen out of the tree and our two dogs were chasing it around like cat and mouse, knocking over furniture and what not. So I think it is safe to say this PR is going out with a BANG! (terrible I know). In case you were wondering, I manage to save the possum from the dogs and relocated it to another tree out the front 👍 🇦🇺

Hahaha you just made my day! This is highly likely the most australian comment this repo has ever seen 😆

Awesome, let's ship this as soon as CI has passed and land it in v3.3!
